src/core/algorithms/pac/pac_verifier/domain_pac_verifier/domain_pac_verifier_base.cpp

```diff
@@ -0,0 +1,153 @@
+                    sorted_value_tuples_.begin(), std::next(first_it),
+                    [this, eps](auto const it) { return domain_->DistFromDomain(*it) > eps; });
+        }
+
+        after_last_it = std::ranges::partition_point(
+                after_last_it, sorted_value_tuples_.end(),
+                [this, eps](auto const it) { return domain_->DistFromDomain(*it) <= eps; });
+        falls_into_domain.push_back(std::distance(first_it, after_last_it));
+    }
+    return falls_into_domain;
+}
+
+unsigned long long DomainPACVerifierBase::ExecuteInternal() {
+    auto start = std::chrono::system_clock::now();
+    std::ostringstream oss;
+    oss << '{';
+    for (auto it = column_indices_.begin(); it != column_indices_.end(); ++it) {
+        if (it != column_indices_.begin()) {
+            oss << ", ";
+        }
+        oss << *it;
+    }
+    oss << '}';
+    LOG_INFO("Verifying Domain PAC on columns {} with domain {}", oss.str(), domain_->ToString());
+    auto satisfying_tuples = CountSatisfyingTuples();
+
+    std::vector<double> empirical_probabilities(satisfying_tuples.size());
+    auto const num_rows = TypedRelation().GetNumRows();
+    std::ranges::transform(satisfying_tuples, empirical_probabilities.begin(),
+                           [num_rows](unsigned long const number) {
+                               return static_cast<double>(number) / num_rows;
+                           });
+    auto [eps, delta] = FindEpsilonDelta(empirical_probabilities);
+
+    auto const schema = TypedRelation().GetSharedPtrSchema();
+    MakePAC<model::DomainPAC>(
+            schema, eps, delta, domain_,
+            schema->GetVertical(util::IndicesToBitset(column_indices_, schema->GetNumColumns())));
+
+    LOG_INFO("Result: {}", GetPAC().ToLongString());
+    auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(
+                           std::chrono::system_clock::now() - start)
+                           .count();
+    LOG_INFO("Validation took {}ms", elapsed);
+    return elapsed;
+}
+
+DomainPACHighlight DomainPACVerifierBase::GetHighlights(double eps_1, double eps_2) const {
+    if (eps_1 < 0) {
+        eps_1 = MinEpsilon();
+    }
+    if (eps_2 < 0) {
+        eps_2 = GetPAC().GetEpsilon();
+    }
+
+    if (eps_2 <= eps_1) {
+        return DomainPACHighlight{tuple_type_, original_value_tuples_, std::vector<TuplesIter>{}};
+    }
+
+    auto first_1_it = std::ranges::partition_point(
+            sorted_value_tuples_.begin(), std::next(first_value_it_),
+            [this, eps_1](auto const it) { return domain_->DistFromDomain(*it) > eps_1; });
+    auto first_2_it = std::ranges::partition_point(
+            sorted_value_tuples_.begin(), first_1_it,
+            [this, eps_2](auto const it) { return domain_->DistFromDomain(*it) > eps_2; });
+    auto after_last_1_it = std::ranges::partition_point(
+            std::prev(after_last_value_it_), sorted_value_tuples_.end(),
```

Comment:
```suggestion
            after_last_value_it_, sorted_value_tuples_.end(),
```
When `after_last_value_it_` points to the first value that doesn't fall into the domain, `std::prev(after_last_value_it_)` would always satisfy the predicate, and thus it would never be returned from `partition_point`. Also `std::prev` results in UB when `after_last_value_it_ == sorted_value_tuples_.begin()`.